File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int dfs(TreeNode* root,int& max_sum){
        
            if(root==NULL){
                    return 0;         
            }
            
            
            int max_left=dfs(root->left,max_sum);
            int max_right=dfs(root->right,max_sum);
        
         
            max_left=max_left>0?max_left:0;
            max_right=max_right>0?max_right:0;
            
            int val=max(max_left,max_right)+root->val;
            
            max_sum=max(max_sum,max_left+max_right+root->val);
                
            return val;
    }
    
    
    
    int maxPathSum(TreeNode* root) {
        
            int max_sum=INT_MIN;
            dfs(root,max_sum);
            return max_sum;
           
        
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def __init__(self):
            self.maxim=float(-inf)
    
    def dfs(self,root):
        
            
            if root==None:
                return 0
        
            val1=self.dfs(root.left)
            val2=self.dfs(root.right)
            ans=val1+val2+root.val
            self.maxim=max(self.maxim,ans)
            val=max(val1,val2)+root.val
            return 0 if val<0 else val
     
    
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
            self.dfs(root)
            return self.maxim
        